### PR TITLE
CM-594: Asynchronous Programming Patterns in .NET

### PR DESCRIPTION
**Pull request by akarshvijayan**

### CODE REVIEW

File: async-csharp/AsynchronousProgrammingPatterns/AsynchronousProgrammingPatterns/Providers/EapUserProvider.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        _operationFinished = ProcessOperationFinished;
+        _userService = new UserService();
+    }
+
+    public User GetUser(int userId) => _userService.GetUser(userId);
+
+    public event EventHandler<GetUserCompletedEventArgs>
+      GetUserCompleted;
+
+    public void GetUserAsync(int userId, object userState)
```
Comment (rbychiang) [inline async-csharp/AsynchronousProgrammingPatterns/AsynchronousProgrammingPatterns/Providers/EapUserProvider.cs:23]: @akarshvijayan I can see the `userState` is always passed as null, so do we actually need this param? I believe it's optional to provide this param. I've mentioned this in the review doc as well, additional params should be part of an overloaded method.

**Review by rbychiang — CHANGES_REQUESTED** [review]

File: async-csharp/AsynchronousProgrammingPatterns/AsynchronousProgrammingPatterns/AsyncProgrammingModelHelper.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
+﻿using AsynchronousProgrammingPatterns.Providers;
+using System.Text;
+
+namespace AsynchronousProgrammingPatterns;
+
+public static class AsyncProgrammingModelHelper
+{
+    public static void FetchAndPrintUser(int userId)
+    {
+
```
[the remark this answers is not in the document]
Comment (akarshvijayan) [reply async-csharp/AsynchronousProgrammingPatterns/AsynchronousProgrammingPatterns/AsyncProgrammingModelHelper.cs]: Done.

File: async-csharp/AsynchronousProgrammingPatterns/AsynchronousProgrammingPatterns/Providers/EapUserProvider.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    public EapUserProvider()
+    {
+        _operationFinished = ProcessOperationFinished;
+        _userService = new UserService();
+    }
+
+    public User GetUser(int userId) => _userService.GetUser(userId);
+
+    public event EventHandler<GetUserCompletedEventArgs>
+      GetUserCompleted;
```
[the remark this answers is not in the document]
Comment (akarshvijayan) [reply async-csharp/AsynchronousProgrammingPatterns/AsynchronousProgrammingPatterns/Providers/EapUserProvider.cs]: Done.

File: async-csharp/AsynchronousProgrammingPatterns/AsynchronousProgrammingPatterns/Providers/EapUserProvider.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+    public void GetUserAsync(int userId, object userState)
+    {
+        AsyncOperation operation = AsyncOperationManager
+         .CreateOperation(userState);
```
[the remark this answers is not in the document]
Comment (akarshvijayan) [reply async-csharp/AsynchronousProgrammingPatterns/AsynchronousProgrammingPatterns/Providers/EapUserProvider.cs]: Done.

File: async-csharp/AsynchronousProgrammingPatterns/AsynchronousProgrammingPatterns/Providers/EapUserProvider.cs (hunk 1 above)
Comment (akarshvijayan) [reply async-csharp/AsynchronousProgrammingPatterns/AsynchronousProgrammingPatterns/Providers/EapUserProvider.cs:24]: Created overload.

**Review by rbychiang — APPROVED** [review]

**Pull request merged**